File: src/python/multithreading/04_race_condition.py
"""
This script demonstrates the concept of a race condition in multithreading.
Multiple threads access and modify a shared global variable `COUNTER` without
any synchronization mechanism, leading to unpredictable results.

Each thread runs a function `foo`, which multiplies the global variable `COUNTER`
by the thread's multiplier in 10 iterations. Due to the race condition, the final
value of `COUNTER` depends on the order and timing of the threads' execution.
"""

from threading import Thread
import time

COUNTER = 1


def foo(multiplier):
    global COUNTER
    for _ in range(10):
        time.sleep(0.1)
        local_counter = COUNTER
        local_counter *= multiplier
        print(f"The COUNTER gets multiplied by {multiplier}")
        COUNTER = local_counter


if __name__ == "__main__":
    threads = []
    for i in range(10):
        threads.append(Thread(target=foo, args=(i + 1,)))

    for thread in threads:
        thread.start()

    for thread in threads:
        thread.join()

    print(f"The final value of COUNTER is: {COUNTER}")

File: src/cpp/multithreading/04_race_condition.cpp
#include <iostream>
#include <thread>
#include <vector>

int COUNTER = 1;

void foo(int multiplier) {

  if (COUNTER < 10) {
    COUNTER *= multiplier;
    std::cout << "The COUNTER get's multiplied by: " << multiplier << std::endl;
  }
}

auto main() -> int {

  std::vector<std::thread> threads;

  threads.reserve(10);
  for (int i = 0; i < 10; i++) {
    threads.push_back(std::thread(foo, i + 1));
  }

  for (auto &t : threads) {
    t.join();
  }

  std::cout << "The final value of COUNTER is: " << COUNTER << std::endl;

  return 0;
}
